### PR TITLE
feat(RotationSystem): Deep Learning Support

### PR DESCRIPTION
**Pull request by 1zun4**
## DeepLearning Rotation System

The successor to https://github.com/CCBlueX/LiquidBounce/pull/5642 - which was far from finished, and unfortunately people tried to use it before I had it ready, so they asked for support and gave feedback on an unfinished pull request.

I am now looking for people who can play and record their combat data - in order to have enough for a ready to use model. I have prepared three models myself within 20 minutes total.

## Features
- [x] Deep Learning Framework
- [x] Download Progressbar
- [x] Rotation Model
- [x] Model Managment Command
- [x] Training Recorder
- [x] Combat Recorder
- [x] Base Model

### For the future
- [ ] Curve Recorder
- [ ] Model Configuration
- [ ] Parameter Configuration
- [ ] Model Metadata
- [ ] Model Load On Select

### Minarai Combat (PTB)
image

No Recording. Imagine playing Practice, over... and over... and over...

### Minarai Trainer (PTP)
image

https://github.com/user-attachments/assets/b0745224-c8e1-4b73-91d9-d6b9471722f2

## Model Managment
- `.models create <name>`
- `.models improve <name>` (!WARNING! This will not filter previously used data, remove it before improving!)
- `.models delete <name>`

image

## Roation Configuration

image

In case our model is not able to keep up with high movement, we can use `Correction` which will act as Aim Assist for the Model. This is very useful as e.g. when only training with PTB-Mode we end up with a lot of low-delta data causing our model to move too slowly. This is NOT an issue with PTP-Mode though. We can also amplify the model output, which has a similar effect and makes it much faster, but it will often aim in the wrong places or start to shake as it tries to correct the wrong output on the next tick.

### Model and Data
[Minarai-Izuna-20min-Training.zip](https://github.com/user-attachments/files/18897442/Minarai-Izuna-20min-Training.zip)

### Video

https://github.com/user-attachments/assets/f89517ff-ad4f-453b-ad6b-5129aa6ba4d1

### In-Depth
As we can see, it sort of combines the concept of **Acceleration** as well as **LazyFlick** (known in Legacy). @mems01 has already done a good job of replicating this behaviour in traditional methods, and I still think it's probably the better solution compared to Deep Learning - but it has to rely on the developer to keep upgrading, updating and maintaining the curves, because as @xkeksi has already pointed out, they always have recognisable patterns. Deep learning is a pattern in itself, but it's a very large and dynamic pattern, and with enough data we can make it behave differently - and with just a little bit more data we can build a whole new pattern without writing a single bit of code.

### Explanation

The problem with this model is that we're relying on the LiquidBounce Aim Point Tracker - which will conflict with our PTB model, which wants to introduce its own point tracking. That's why it started to drift a bit - but overall this clip shows how it works, and from the current feedback I'm getting, it seems I've finally got it right.

### Feedback
image
image
image

Thank you very much. It's always nice to hear that something you've been working on for the last 5 days is working. Of course we still have to see how it works with anti-cheats, but I will leave that to the config creators and just provide an easy to use model system :)

## To reviewers
Do NOT submit any code feedback. Thank you.

### CODE REVIEW

**Comment by twcult** [comment]
Hurttime may be superfluous in this case, but I could be wrong. Look at its priority in features

**Comment by 1zun4** [comment]
> Hurttime may be superfluous in this case, but I could be wrong. Look at its priority in features

The data is being recorded - however, it's not passed into the model.

https://github.com/CCBlueX/LiquidBounce/blob/e445ca915fae368523b114410496ba82ed925b46/src/main/kotlin/net/ccbluex/liquidbounce/deeplearn/data/TrainingData.kt#L69-L90

**Review by sqlerrorthing — COMMENTED** [review]
aww fuck izuna fuck me

**Comment by sqlerrorthing** [comment]
@1zun4 do u know russian? (at least a little?)

**Comment by vanvac** [comment]
dam ai aimbot is cazy (anticheats are cooked)

**Comment by sqlerrorthing** [comment]
image
invalid char in progress bar (after percents and before progress bar bar)

**Comment by 1zun4** [comment]
> image
> invalid char in progress bar (after percents and before progress bar bar)

That is normal. It's a line break and only works using the LiquidBounce HUD.

**Comment by sqlerrorthing** [comment]
лол хфывфхывхыфвх 


> > image
> > invalid char in progress bar (after percents and before progress bar bar)
> 
> That is normal. It's a line break and only works using the LiquidBounce HUD.

check for it?

**Comment by 1zun4** [comment]
> check for it?

Not relevant.

File: build.gradle [hunk 1]
```diff
@@ -160,6 +160,13 @@ dependencies {
 //    includeDependency "org.graalvm.polyglot:ruby-community:24.0.2"
 //    includeDependency "org.graalvm.polyglot:llvm-native-community:24.0.2"
 
+    // Machine Learning
+    includeDependency 'ai.djl:api:0.31.1'
+    includeDependency "ai.djl.pytorch:pytorch-engine:0.31.1"
+//    runtimeOnly "ai.djl.mxnet:mxnet-engine:0.31.1"
+//    runtimeOnly "ai.djl.tensorflow:tensorflow-model-zoo:0.31.1"
```
Comment (superblaubeere27) [inline build.gradle:169]: image

**Review by superblaubeere27 — DISMISSED** [review]

File: build.gradle (hunk 1 above)
Comment (sqlerrorthing) [reply build.gradle:169]: do not review a draft pr

**Comment by commandblock2** [comment]
Holy *, I am speechless, never thought a tiny model like this could be perfect for this task
```
private fun createMlpBlock(outputs: Long) = SequentialBlock()
    .add(Linear.builder()
        .setUnits(128)
        .build())
    .add(Blocks.batchFlattenBlock())
    .add(BatchNorm.builder().build())
    .add(Activation.reluBlock())

    .add(Linear.builder()
        .setUnits(64)
        .build())
    .add(Blocks.batchFlattenBlock())
    .add(BatchNorm.builder().build())
    .add(Activation.reluBlock())

    .add(Linear.builder()
        .setUnits(32)
        .build())
    .add(Blocks.batchFlattenBlock())
    .add(BatchNorm.builder().build())
    .add(Activation.reluBlock())

    .add(Linear.builder()
        .setUnits(outputs)
        .build())
```

You did make it happen and that's crazy. I think this is the exactly right step to go.
I am indeed thinking about a neural network in the game tho I would do it with the client sending gamedata to / receiving input from another process (python or whatever).

**Comment by sqlerrorthing** [comment]
image
FUCK
AN HOUR OF RECORDING THIS SHIT

**Comment by sqlerrorthing** [comment]
@1zun4, record with 100k samples (again, but it was saved successfully this time.)
download [record (55.4 MiB)](https://pomf2.lain.la/f/1yc28l82.json), [trained model (45.5 KiB)](https://github.com/user-attachments/files/18921296/sqle-100k.zip)

**Comment by MukjepScarlet** [comment]
@sqlerrorthing we probably need a more effecient way to save the record XD

**Comment by MukjepScarlet** [comment]
Well I found a eaiser way to solve the save issue

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/misc/debugrecorder/ModuleDebugRecorder.kt [hunk 2]
```diff
@@ -43,31 +47,24 @@ object ModuleDebugRecorder : ClientModule("DebugRecorder", Category.MISC) {
...
-
                 var file = folder.resolve("${baseName}.json")
 
                 var idx = 0
-
                 while (file.exists()) {
                     file = folder.resolve("${baseName}_${idx++}.json")
                 }
 
                 file.writeText(publicGson.toJson(this.packets))
```
Comment (MukjepScarlet) [inline src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/misc/debugrecorder/ModuleDebugRecorder.kt]: This can fix the save failure:
```kotlin
file.bufferedWriter().use {
    publicGson.toJson(this.packets, it)
}
```

Comment (sqlerrorthing) [reply src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/misc/debugrecorder/ModuleDebugRecorder.kt]: do not review, and check [it](https://github.com/CCBlueX/LiquidBounce/pull/5672/commits/e81196c4dc71614dc8d99980b090827276a45306#diff-84d2bd893025a299a7a199637be011492db5c859cab6afc92ef8980c5133e573R73-R76)

Comment (MukjepScarlet) [reply src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/misc/debugrecorder/ModuleDebugRecorder.kt]: Well that's another problem, we should make sure the `folder` exists before saving

Comment (sqlerrorthing) [reply src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/misc/debugrecorder/ModuleDebugRecorder.kt]: we sure.
https://github.com/CCBlueX/LiquidBounce/blob/e81196c4dc71614dc8d99980b090827276a45306/src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/misc/debugrecorder/ModuleDebugRecorder.kt#L71

**Comment by 1zun4** [comment]
> @1zun4, record with 100k samples (again, but it was saved successfully this time.) download [record (55.4 MiB)](https://pomf2.lain.la/f/1yc28l82.json), [trained model (45.5 KiB)](https://github.com/user-attachments/files/18921296/sqle-100k.zip)

Thank you. That is with PTP Training (Minarai Trainer)?

**Comment by sqlerrorthing** [comment]
> > @1zun4, record with 100k samples (again, but it was saved successfully this time.) download [record (55.4 MiB)](https://pomf2.lain.la/f/1yc28l82.json), [trained model (45.5 KiB)](https://github.com/user-attachments/files/18921296/sqle-100k.zip)
> 
> Thank you. That is with PTP Training (Minarai Trainer)?

yes

**Comment by MukjepScarlet** [comment]
It will be good if we can share and get models on the incoming marketplace, I'm looking forward to it.

**Comment by vanvac** [comment]
will this feature require a decent computer to run cause from what im seeing its a ai controlling the player and ai dont seem too easy to run on not good pc's

**Comment by sqlerrorthing** [comment]
> will this feature require a decent computer to run cause from what im seeing its a ai controlling the player and ai dont seem too easy to run on not good pc's

i think not. maybe add a setting, like a toggle switch.
remember that ai doesn't need many neurons to work

**Comment by liquidsquid1** [comment]
this seems really interesting! i can do a 6 hour recording soon and upload the model here with minarai combat? ill just be playing pika xd

**Comment by DrParanoya** [comment]
It looks way too legit

**Comment by vanvac** [comment]
is this planned to release with lb 1.27.0 or its done when its done type of thing

**Comment by 1zun4** [comment]
> @1zun4, record with 100k samples (again, but it was saved successfully this time.) download [record (55.4 MiB)](https://pomf2.lain.la/f/1yc28l82.json), [trained model (45.5 KiB)](https://github.com/user-attachments/files/18921296/sqle-100k.zip)

image
Next time please upload to GitHub instead. Also compress the JSON before uploading.

**Comment by sqlerrorthing** [comment]
> > @1zun4, record with 100k samples (again, but it was saved successfully this time.) download [record (55.4 MiB)](https://pomf2.lain.la/f/1yc28l82.json), [trained model (45.5 KiB)](https://github.com/user-attachments/files/18921296/sqle-100k.zip)
> 
> image Next time please upload to GitHub instead. Also compress the JSON before uploading.

oh, but this file more than 10 MiB

**Comment by 1zun4** [comment]
> > > @1zun4, record with 100k samples (again, but it was saved successfully this time.) download [record (55.4 MiB)](https://pomf2.lain.la/f/1yc28l82.json), [trained model (45.5 KiB)](https://github.com/user-attachments/files/18921296/sqle-100k.zip)
> > 
> > 
> > image Next time please upload to GitHub instead. Also compress the JSON before uploading.
> 
> oh, but this file more than 10 MiB

Not as ZIP.
[1yc28l82.zip](https://github.com/user-attachments/files/18951761/1yc28l82.zip) (4,1 MiB)

**Comment by MukjepScarlet** [comment]
We can have a special file format for `List<TrainingData>`. But this will make the files unreadable.

**Comment by sqlerrorthing** [comment]
> > > > @1zun4, record with 100k samples (again, but it was saved successfully this time.) download [record (55.4 MiB)](https://pomf2.lain.la/f/1yc28l82.json), [trained model (45.5 KiB)](https://github.com/user-attachments/files/18921296/sqle-100k.zip)
> > > 
> > > 
> > > image Next time please upload to GitHub instead. Also compress the JSON before uploading.
> > 
> > oh, but this file more than 10 MiB
> 
> Not as ZIP.
> [1yc28l82.zip](https://github.com/user-attachments/files/18951761/1yc28l82.zip) (4,1 MiB)
> 
> 

не подумал, сорянчик

**Comment by 1zun4** [comment]
> We can have a special file format for `List<TrainingData>`. But this will make the files unreadable.

That is not an issue but compressed JSON does the job as well.

**Comment by xkeksi** [comment]
: )

**Comment by liquidsquid1** [comment]
> this seems really interesting! i can do a 6 hour recording soon and upload the model here with minarai combat? ill just be playing pika xd

haven't been banned yet, wow this does really look realistic.

**Comment by 1zun4** [comment]
Just to let you know. This pull request is basically done, but I need to work on a new startup task manager that notifies the user that a required file is being downloaded, at what speed, and how much % is completed, similar to what we already do with MCEF - but this time integrated into LiquidBounce instead. At the same time, we can also use this to speed up the startup time of LiquidBounce in the future, as I will be working on moving more of the client initialization to asynchronous. Fortunately, we can inject into https://github.com/deepjavalibrary/djl/blob/master/api/src/main/java/ai/djl/util/Utils.java using a mixin - this allows us to monitor the progress of a DJL download. Currently DJL just prints it to the console without providing access to a progress listener, so we need to integrate one ourselves.

I think I will do this in another pull request though, and when that is done, integrate it properly here.

**Comment by sqlerrorthing** [comment]
separate the library's downloading menu in mcef, provide only api for download & the download progress.

create downloading tasks, download the mcef library & download the deeplearn library.

create the downloading menu in the client and asynchronous run tasks & render their progress

i see it

File: build.gradle (hunk 1 above)
Comment (superblaubeere27) [reply build.gradle:169]: > do not review a draft pr

that's... the joke

**Comment by 7GrandDadPGN** [comment]
3 hours of playing on gamster practice with no mitigation, the deep learning bypass go crazy 🗣️🗣️
(I did have to add my own sword hand check to aura though to play legit, but it was fun)
image

**Comment by 24dedlaa** [comment]
So how do I turn the recorded json from the debug recorder module into a model?

**Comment by sqlerrorthing** [comment]
> So how do I turn the recorded json from the debug recorder module into a model?

put your <file>.json into:
`<minecraft dir>/LiquidBounce/debug-recorder/MinaraiTrainer/<file>.json`

**Comment by 24dedlaa** [comment]
> > So how do I turn the recorded json from the debug recorder module into a model?
> 
> put your .json into: `<minecraft dir>/LiquidBounce/debug-recorder/MinaraiTrainer/<file>.json`

Ok, I made my model, played for a few minutes with debug recorder enabled with the minaraicombat mode. It wrote me where the model was saved as json. Now how do I make it so I can select it in killaura?

**Comment by SuicidalEeper** [comment]
> > > So how do I turn the recorded json from the debug recorder module into a model?
> > 
> > 
> > put your .json into: `<minecraft dir>/LiquidBounce/debug-recorder/MinaraiTrainer/<file>.json`
> 
> Ok, I made my model, played for a few minutes with debug recorder enabled with the minaraicombat mode. It wrote me where the model was saved as json. Now how do I make it so I can select it in killaura?

.models create name then  it's under rotation modes and you need both minaraicombat and minaraitrainer with like 10000 samples each

**Comment by 1zun4** [comment]
75K Model with Sigmoid Correction Test
https://www.youtube.com/watch?v=6WbdAtx_l5w

**Comment by CaptainBoggle** [comment]
@1zun4 Sorry for the ping, just had an idea and was wondering if it seems viable/makes sense at all to you:
Instead of recording combat data of the current player, could we instead record data from other players (even multiple simultaneously), such that we could simply AFK in spectator mode on a high-tier pvp server rather than relying on either the pretrained model (which could potentially be detectable at some point) or the users own pvp skill lol. This would also give us a lot more data to play with in a reasonable amount of time, as you could have alts recording across many servers/lobbies. The data won't be as precise due to the limitations of needing to interpret packets, and training on server-processed movements could result in some potentially strange behaviour such as the client never sending slightly out of bounds movements which the server would normally clamp etc. Anyway, just wanted to see if you had already considered this.

**Comment by 1zun4** [comment]
> @1zun4 Sorry for the ping, just had an idea and was wondering if it seems viable/makes sense at all to you: Instead of recording combat data of the current player, could we instead record data from other players (even multiple simultaneously), such that we could simply AFK in spectator mode on a high-tier pvp server rather than relying on either the pretrained model (which could potentially be detectable at some point) or the users own pvp skill lol. This would also give us a lot more data to play with in a reasonable amount of time, as you could have alts recording across many servers/lobbies. The data won't be as precise due to the limitations of needing to interpret packets, and training on server-processed movements could result in some potentially strange behaviour such as the client never sending slightly out of bounds movements which the server would normally clamp etc. Anyway, just wanted to see if you had already considered this.

I have already tried it but it is a bit more problematic, because we basically need to detect if someone is in combat. If we start having samples where e.g. someone is only running away instead of being in combat, the model would be badly impacted. Also it would be better if we are the server instead of trying to record data from other players while being in client. I tried this on Hypixel Pit and did not manage to find a good way to do that.

With a lot of time you could of course do this, but at the moment it's much easier to record the combat data yourself.
